skimage/registration/_lddmm.py

```diff
@@ -0,0 +1,1247 @@
+        affine_inv_hessian_approx = np.sum(affine_inv_hessian_approx, tuple(range(self.target.ndim)))
+
+        # Solve for affine_inv_gradient.
+        try:
+            affine_inv_gradient = solve(affine_inv_hessian_approx, affine_inv_gradient_reduction, assume_a='pos').reshape(matching_affine_inv_gradient.shape[-2:])
+        except np.linalg.LinAlgError as exception:
+            exception.args = (
+                *exception.args, f"The Hessian was not invertible in the Gauss-Newton update of the affine transform. "
+                                 f"This may be because the image was constant along one or more dimensions. "
+                                 f"Consider removing any constant dimensions. "
+                                 f"Otherwise you may try using a smaller value for affine_stepsize, a smaller value for deformative_stepsize, or a larger value for sigma_regularization. "
+                                 f"The calibrate=True option may be of use in determining optimal parameter values."
+            )
+            raise exception
+        # Append a row of zeros at the end of the 0th dimension.
+        zeros = np.zeros((1, self.target.ndim + 1))
+        affine_inv_gradient = np.concatenate((affine_inv_gradient, zeros), 0)
+
+        return affine_inv_gradient
+
+
+    def _update_affine(self, affine_inv_gradient, iteration):
+        """
+        Update self.affine based on affine_inv_gradient.
+
+        If iteration < self.num_rigid_affine_iterations, project self.affine to a rigid affine.
+
+        If self.
+         is provided, it is imposed on self.affine.
+
+        if self.calibrate, appends the current self.affine to self.affines.
+
+        Updates attributes:
+            affine
+            affines
+        """
+        
+        affine_inv = inv(self.affine)
+
+        affine_inv -= affine_inv_gradient * self.affine_stepsize
+
+        self.affine = inv(affine_inv)
+
+        # Set scale of self.affine if appropriate.
+        if self.fixed_affine_scale is not None:
+            U, _, Vh = svd(self.affine[:-1, :-1])
+            self.affine[:-1, :-1] = U @ np.diag([self.fixed_affine_scale] * len(self.affine) - 1) @ Vh
+        # If self.fixed_affine_scale was not provided (is None), project self.affine to a rigid affine if appropriate.
+        elif iteration < self.num_rigid_affine_iterations:
+            U, _, Vh = svd(self.affine[:-1, :-1])
+            self.affine[:-1, :-1] = U @ Vh
+
+        # Save affine for calibration plotting.
+        if self.calibrate:
+            self.affines.append(self.affine)
+
+
+    def _compute_velocity_fields_gradients(self):
+        """
+        Compute and return the gradients of the self.velocity_fields.
+
+        Updates attributes:
+            phi
+            affine_phi
+        """
+
+        matching_error_prime = (self.contrast_deformed_template - self.target) * self.matching_weights / self.sigma_matching**2
```

Comment:
I agree that the use of a class here is essentially to allow access to many global variables internally. Each method is called from the central register method and they otherwise do not call each other. Thus, it could be readily converted to a purely functional style.

My opinion on it is that their functional separation slashes the cost to comprehensibility for being methods, and due to the number of parameters used and updated across these methods, if they were functions the calls and signatures would be the gnarliest things since the tree from Ferngully.